Migrate to AWS SDKv2

torredil opened the pull request:
**Is this a bug fix or adding new feature?**

Dependency migration

**What is this PR about? / Why do we need it?**

In alignment with our [SDKs and Tools Maintenance Policy](https://docs.aws.amazon.com/sdkref/latest/guide/maint-policy.html), [AWS SDK for Go (v1)](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/index.html) will enter maintenance mode on July 31, 2024 and reach end-of-support on July 31, 2025. This PR implements the necessary changes to migrate to AWS SDK for Go v2.

Fixes #1684

**Additional Information**

https://aws.github.io/aws-sdk-go-v2/docs/migrating/

torredil commented:
/hold

torredil commented:
/retest

AndrewSirenko reviewed (commented):
Overall thank you so much for doing a great job on this herculean migration! 

I still need to review a few more files, and run some more manual tests, but overall this looks great so far. 

Have a gift of a few questions and nits to for now for your efforts 😅. 

Files left to review:

```
cloud_test.go
handlers.go
metadata_test.go
controller.go
controller_test.go
```

AndrewSirenko commented:
/retest

torredil commented on pkg/cloud/metadata_ec2.go:
```diff
@@ -1,31 +1,46 @@
 
 func EC2MetadataInstanceInfo(svc EC2Metadata, regionFromSession string) (*Metadata, error) {
-	doc, err := svc.GetInstanceIdentityDocument()
-	klog.InfoS("Retrieving EC2 instance identity metadata", "regionFromSession", regionFromSession)
```
The equivalent of this informational is logged in `func NewMetadataService`

torredil commented on pkg/cloud/metadata_ec2.go:
```diff
@@ -80,21 +98,18 @@ func EC2MetadataInstanceInfo(svc EC2Metadata, regionFromSession string) (*Metada
-	outpostArn, err := svc.GetMetadata(outpostArnEndpoint)
-	// "outpust-arn" returns 404 for non-outpost instances. note that the request is made to a link-local address.
-	// it's guaranteed to be in the form `arn:<partition>:outposts:<region>:<account>:outpost/<outpost-id>`
-	// There's a case to be made here to ignore the error so a failure here wouldn't affect non-outpost calls.
```
Added explicit 404 check

torredil commented on pkg/cloud/cloud.go:
```diff
@@ -639,15 +631,16 @@ func (c *cloud) CreateDisk(ctx context.Context, volumeName string, diskOptions *
 
 	newSizeGiB := util.RoundUpGiB(newSizeBytes)
-	needsModification, volumeSize, err := c.validateModifyVolume(ctx, volumeID, newSizeGiB, options)
+	needsModification, volumeSize, err := c.validateModifyVolume(ctx, volumeID, int32(newSizeGiB), options)
```
Done

torredil commented on pkg/cloud/cloud.go:
```diff
@@ -940,10 +929,11 @@ func (c *cloud) WaitForAttachmentState(ctx context.Context, volumeID, expectedSt
-					attachmentState = aws.StringValue(a.State)
-					attachment = a
+			a := a
+			if a.InstanceId != nil {
```
Yes - `a.State` is always a valid value within the `types.VolumeAttachment` struct.

torredil commented on pkg/cloud/cloud.go:
```diff
@@ -1070,15 +1062,16 @@ func execBatchDescribeSnapshots(svc ec2iface.EC2API, input []string, batcher sna
 
 	for _, snapshot := range resp {
-		key, err := extractSnapshotKey(snapshot, batcher)
+		snapshot := snapshot
```
The linter is still flagging it, golangci-lint has not yet been updated.

torredil commented on pkg/cloud/cloud.go:
```diff
@@ -1087,24 +1080,24 @@ func execBatchDescribeSnapshots(svc ec2iface.EC2API, input []string, batcher sna
 
 	switch {
-	case len(request.SnapshotIds) == 1 && request.SnapshotIds[0] != nil:
+	case len(request.SnapshotIds) == 1:
```
Added the check here as well

torredil commented on pkg/cloud/cloud.go:
```diff
@@ -1146,39 +1139,42 @@ func extractSnapshotKey(s *ec2.Snapshot, batcher snapshotBatcherType) (string, e
 		return nil, fmt.Errorf("nil CreateSnapshotResponse")
 	}
 
-	return c.ec2SnapshotResponseToStruct(res), nil
```
`CreateSnapshot` returns `*ec2.CreateSnapshotOutput`. Other snapshot functions such as `GetSnapshotByName` return `*types.Snapshot`.

torredil commented on pkg/cloud/cloud.go:
```diff
@@ -1551,24 +1537,25 @@ func isAWSErrorBlockDeviceInUse(err error) bool {
+	if errors.As(err, &apiError) {
+		msg := apiError.Error()
+		switch {
+		case strings.Contains(msg, "InvalidParameter"):
```
Yes can do! updated function in new rev is much cleaner

torredil commented on pkg/cloud/cloud.go:
```diff
@@ -1762,33 +1750,30 @@ func (c *cloud) validateModifyVolume(ctx context.Context, volumeID string, newSi
 	for _, attachment := range volume.Attachments {
-		if attachment.State != nil && strings.ToLower(aws.StringValue(attachment.State)) == volumeAttachedState {
-			volumeAttachmentList = append(volumeAttachmentList, aws.StringValue(attachment.InstanceId))
+		if attachment.State == volumeAttachedState {
```
Yes, as per the SDK implementation `attachment.State` value is always valid within the `types.VolumeAttachment` struct.

torredil commented:
Fixed `CreateSnapshot` unit test - turns out, `tc.expInput` is created in the test function while the actual input passed to `CreateSnapshot` is created inside the func, even though the contents of the structs are the same they are separate instances with different memory addresses and hence why the equality check failed.

Fixed now + the go sum merge conflict. No issues spotted in manual testing + external e2e tests are all green.

The last remaining action item is RCAing why `make verify` hangs here in CI environment, but succeeds locally:

```
$ make verify
go vet $(go list ./...)
INFO Execution took 960.184531ms
./hack/verify-update.sh
All verifications passed!
```

torredil commented:
Upgrading linter for `make verify` in https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/1971

torredil commented:
Logs captured during manual testing for new middleware handler implementation used to record AWS API metrics:
```
Handling connection for 3301
# HELP cloudprovider_aws_api_request_duration_seconds [ALPHA] ebs_csi_aws_com metric
# TYPE cloudprovider_aws_api_request_duration_seconds histogram
cloudprovider_aws_api_request_duration_seconds_bucket{request="AttachVolume",le="0.005"} 0
cloudprovider_aws_api_request_duration_seconds_bucket{request="AttachVolume",le="0.01"} 0
cloudprovider_aws_api_request_duration_seconds_bucket{request="AttachVolume",le="0.025"} 0
cloudprovider_aws_api_request_duration_seconds_bucket{request="AttachVolume",le="0.05"} 0
cloudprovider_aws_api_request_duration_seconds_bucket{request="AttachVolume",le="0.1"} 0
cloudprovider_aws_api_request_duration_seconds_bucket{request="AttachVolume",le="0.25"} 0
cloudprovider_aws_api_request_duration_seconds_bucket{request="AttachVolume",le="0.5"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="AttachVolume",le="1"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="AttachVolume",le="2.5"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="AttachVolume",le="5"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="AttachVolume",le="10"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="AttachVolume",le="+Inf"} 1
cloudprovider_aws_api_request_duration_seconds_sum{request="AttachVolume"} 0.386986709
cloudprovider_aws_api_request_duration_seconds_count{request="AttachVolume"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="CreateVolume",le="0.005"} 0
cloudprovider_aws_api_request_duration_seconds_bucket{request="CreateVolume",le="0.01"} 0
cloudprovider_aws_api_request_duration_seconds_bucket{request="CreateVolume",le="0.025"} 0
cloudprovider_aws_api_request_duration_seconds_bucket{request="CreateVolume",le="0.05"} 0
cloudprovider_aws_api_request_duration_seconds_bucket{request="CreateVolume",le="0.1"} 0
cloudprovider_aws_api_request_duration_seconds_bucket{request="CreateVolume",le="0.25"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="CreateVolume",le="0.5"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="CreateVolume",le="1"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="CreateVolume",le="2.5"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="CreateVolume",le="5"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="CreateVolume",le="10"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="CreateVolume",le="+Inf"} 1
cloudprovider_aws_api_request_duration_seconds_sum{request="CreateVolume"} 0.244747201
cloudprovider_aws_api_request_duration_seconds_count{request="CreateVolume"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="DescribeInstances",le="0.005"} 0
cloudprovider_aws_api_request_duration_seconds_bucket{request="DescribeInstances",le="0.01"} 0
cloudprovider_aws_api_request_duration_seconds_bucket{request="DescribeInstances",le="0.025"} 0
cloudprovider_aws_api_request_duration_seconds_bucket{request="DescribeInstances",le="0.05"} 0
cloudprovider_aws_api_request_duration_seconds_bucket{request="DescribeInstances",le="0.1"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="DescribeInstances",le="0.25"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="DescribeInstances",le="0.5"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="DescribeInstances",le="1"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="DescribeInstances",le="2.5"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="DescribeInstances",le="5"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="DescribeInstances",le="10"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="DescribeInstances",le="+Inf"} 1
cloudprovider_aws_api_request_duration_seconds_sum{request="DescribeInstances"} 0.067674695
cloudprovider_aws_api_request_duration_seconds_count{request="DescribeInstances"} 1
cloudprovider_aws_api_request_duration_seconds_bucket{request="DescribeVolumes",le="0.005"} 0
cloudprovider_aws_api_request_duration_seconds_bucket{request="DescribeVolumes",le="0.01"} 0
```

/unhold

ConnorJC3 commented on pkg/cloud/cloud_test.go:
```diff
@@ -1711,28 +1745,39 @@ func TestCreateSnapshot(t *testing.T) {
-			mockEC2.EXPECT().CreateSnapshotWithContext(gomock.Any(), eqCreateSnapshotInput(tc.expInput)).Return(ec2snapshot, tc.expErr)
-			mockEC2.EXPECT().DescribeSnapshotsWithContext(gomock.Any(), gomock.Any()).Return(&ec2.DescribeSnapshotsOutput{Snapshots: []*ec2.Snapshot{ec2snapshot}}, nil).AnyTimes()
+
+			mockEC2.EXPECT().CreateSnapshot(gomock.Any(), gomock.Any(), gomock.Any()).Return(&ec2.CreateSnapshotOutput{
```
If we're going to remove `eqCreateSnapshotInput`, is it worth using `DoAndReturn` to check the AWS SDK call contains the correct volume id/tags/description?

torredil replied:
Sure, added `DoAndReturn` validation.

ConnorJC3 reviewed (commented):
Initial review - very little manual testing

Intend to do the bulk of manual testing early next week, after which I will publish a final review

(expect very little additional comments in the final review unless I catch a bug with the manual testing)

AndrewSirenko commented on pkg/cloud/cloud.go:
```diff
@@ -1694,38 +1669,43 @@ func (c *cloud) AvailabilityZones(ctx context.Context) (map[string]struct{}, err
 	}
-	if options.VolumeType != "" && !strings.EqualFold(*volume.VolumeType, options.VolumeType) {
+
+	if options.VolumeType != "" && string(volume.VolumeType) != strings.ToLower(options.VolumeType) {
```
Is it fine to remove at line (old) 1603 but not here?

ConnorJC3 replied:
Preferably keep there too

torredil replied:
Done ✅

torredil commented on pkg/driver/controller.go:
```diff
@@ -162,22 +168,25 @@ func (d *controllerService) CreateVolume(ctx context.Context, req *csi.CreateVol
 		case IopsPerGBKey:
-			iopsPerGB, err = strconv.Atoi(value)
-			if err != nil {
+			parseIopsPerGBKey, parseIopsPerGBKeyErr := strconv.Atoi(value)
```
Done ✅

torredil commented on tests/e2e/testsuites/e2e_utils.go:
```diff
@@ -68,7 +68,7 @@ func PodCmdGrepVolumeData(volumeMountPath string) string {
 	pvcSize := pvc.Spec.Resources.Requests["storage"]
 	delta := resource.Quantity{}
-	delta.Set(util.GiBToBytes(sizeIncreaseGi))
+	delta.Set(util.GiBToBytes(int32(sizeIncreaseGi)))
```
Changing the parameter to `int32` is cleaner, made the change.

torredil commented on pkg/util/util.go:
```diff
@@ -28,33 +29,42 @@ import (
+	result := roundUpSize(volumeSizeBytes, GiB)
+	if result > int64(math.MaxInt32) {
+		// Handle overflow
+		return math.MaxInt32
```
Good idea - best to bubble up the error instead of silently capping to give the caller a chance to handle the overflow.

torredil commented on pkg/driver/controller_test.go:
```diff
@@ -47,107 +47,6 @@ const (
 	expDevicePath = "/dev/xvda"
 )
 
-func TestNewControllerService(t *testing.T) {
```
Because this unit test is a disaster 😄 

To give a bit more context here, the previous implementation relied on global variables to override their assignment for the single purpose of unit testing:

```
var (
	// NewMetadataFunc is a variable for the cloud.NewMetadata function that can
	// be overwritten in unit tests.
	NewMetadataFunc = cloud.NewMetadataService
	// NewCloudFunc is a variable for the cloud.NewCloud function that can
	// be overwritten in unit tests.
	NewCloudFunc = cloud.NewCloud
)
```

metadata and cloud are created within `newControllerService` which makes unit testing its functionality incredibly difficult. 

Fixing the relationship between cloud, controller, and metadata is beyond the scope of this PR, but I will be addressing that in an upcoming PR to fix this pain point. For now, I've rolled back metadata instantiation in controller `cloud.NewMetadataService(cfg, region)` -> `NewMetadataFunc(cfg, region)` to re-implement unit test.

torredil commented on pkg/cloud/cloud.go:
```diff
@@ -1551,24 +1537,17 @@ func isAWSErrorBlockDeviceInUse(err error) bool {
+			"UnknownParameter",
+			"UnknownVolumeType",
+			"UnsupportedOperation",
+			"ValidationError":
```
Good idea - makes the code more concise and readable.

torredil commented on pkg/cloud/metadata_test.go:
```diff
@@ -1,435 +1,608 @@
 You may obtain a copy of the License at
 
-    http://www.apache.org/licenses/LICENSE-2.0
+	http://www.apache.org/licenses/LICENSE-2.0
```
Done ✅

torredil commented on pkg/cloud/metadata_test.go:
```diff
@@ -1,435 +1,608 @@
 See the License for the specific language governing permissions and
 limitations under the License.
 */
-
```
Done ✅

ConnorJC3 commented:
/lgtm

I tried my hardest to break it, and I failed.

AndrewSirenko commented on pkg/cloud/cloud.go:
```diff
@@ -280,29 +289,27 @@ func NewCloud(region string, awsSdkDebugLog bool, userAgentExtra string, batchin
-		CredentialsChainVerboseErrors: aws.Bool(true),
-		// Set MaxRetries to a high value. It will be "overwritten" if context deadline comes sooner.
-		MaxRetries: aws.Int(8),
+	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion(region))
```
Nit: It may technically be out of scope for this PR, but I don't think we shouldhave `context.TODO()`here  in our next release (perhaps a high priority backlog item?)

And maybe we can rely on this future context as the timeout for other driver-startup tasks such as taint removal? To have one source of truth for `driver startup is taking too long, lets restart the container` rather than our amalgamation of different timeouts?

torredil replied:
Switched from `context.TODO()` to `context.Background()` given a timeout here is not incredibly relevant since any failures at this point will be handled by the livenessprobe.

AndrewSirenko reviewed (approved):
Great work on this tricky migration. Thank you for delivering results.

torredil commented:
/approve
cc: @ConnorJC3 @AndrewSirenko

ConnorJC3 commented:
/lgtm
